04-onion-routing.md

```diff
@@ -367,9 +378,60 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Trampoline Payments
+
+Trampoline payments allow nodes with an incomplete view of the network to
+delegate the construction of parts of the route to trampoline nodes.
+
+The origin node only needs to select a set of trampoline nodes and to know a
+route to the first trampoline node. Each trampoline node is responsible for
+finding its own route to the next trampoline node. The last trampoline node
+must be the final recipient.
+
+The `trampoline_onion_packet` has a variable size to allow implementations to
+choose their own trade-off between flexibility and privacy. It's recommended to
+add trailing filler data to the `trampoline_onion_packet` when using a small
+number of hops. It uses the same onion construction as the `onion_packet` and
+is embedded inside an `onion_packet`.
+
+Trampoline nodes are free to use as many hops as they want between themselves
+as long as they are able to create a route that satisfies the `cltv` and `fees`
+requirements contained in the onion.
+
+#### Requirements
+
+A sending node:
+
+- If the invoice doesn't support the `trampoline_routing` feature:
+  - MUST NOT use trampoline routing to pay that invoice
+- MUST ensure that each hop in the `trampoline_onion_packet` supports `trampoline_routing`
+- MUST encrypt the `trampoline_onion_packet` with the same construction as `onion_packet`
+- MAY add trailing filler data similar to what is done in the `onion_packet`
+- MUST use a different `session_key` for the `trampoline_onion_packet` and the `onion_packet`
+- MUST include the `trampoline_onion_packet` tlv in the _last_ hop's payload of the `onion_packet`
+- MUST include the invoice's `payment_secret` in the _last_ hop's payload of the `trampoline_onion_packet`
+- MUST generate a different `payment_secret` to use in the outer onion
+
+When processing a `trampoline_onion_packet`, a receiving node:
+
+- If it doesn't support `trampoline_routing`:
+  - MUST report a route failure to the origin node
+- Otherwise, if it supports `trampoline_routing`:
+  - MUST process the `trampoline_onion_packet` as an `onion_packet`
+  - MUST fail the HTLC if dictated by the requirements under [Failure Messages](#failure-messages)
+  - MUST compute a route to the next trampoline node
+  - MUST include the peeled `trampoline_onion_packet` in the `hop_payload` for the next trampoline node
+  - MAY aggregate the incoming multi-part payment before forwarding
```

Comment:
Oh, that's interesting!  Not sure we'll be implementing that for v1 ourselves though...

Reply:
Definitely something that can be done in a second step!

But since you'll be calculating routes anyway even to relay single-part HTLCs, you'll see that it won't be much additional work to aggregate all the incoming parts before relaying.

One point that was a bit hard to get right in eclair is the HTLC reconciliation after a restart (if for example the node was restarted before an HTLC was relayed out), because this breaks the mapping "1 HTLC out <-> 1 HTLC in", it's now "M HTLCs out <-> N HTLCs in".

Reply:
If the receiving nodes aggregates the MPP, then I think there should be a requirement to fail the payment if the `trampoline_onion_packet` is not the same in every part of the MPP. I recently found that Eclair is not doing that.

I think this also implies that the decision to aggregate the payment should not be left with the trampoline node. If the receiver of the MPP is the trampoline node, I do not see how it could decide not to aggregate the payment.

Reply:
> I think there should be a requirement to fail the payment if the trampoline_onion_packet is not the same in every part of the MPP

Yes that would make sense to add an explicit requirement here.
For now eclair is quite lax here (which lets the final recipient decide whether the payment is valid or not), but that flexibility may not bring anything to the table, so it would make sense to reject that kind of HTLCs, I'll look into it.

> I think this also implies that the decision to aggregate the payment should not be left with the trampoline node.

I think it does, as we cannot force an implementation to do something if it's not related to security.
A trampoline node may chose to simply forward each HTLC independently without aggregating them if it doesn't have the resources to aggregate (which requires allocating resources to hodl parts until the complete payment arrives).
This is a different requirement from the case where the trampoline node is the final recipient.

Reply:
Oh I see. That distinction was missing, I think it should be clarified:
 - if the trampoline node is the final recipient of the MPP, it must check that `trampoline_onion_packets` in each part are identical before forwarding
 - if the trampoline node is not the final recipient of the MPP, it may aggregate HTLCs before forwarding, but what does "aggregate" mean in that context? simply to wait? what if the sender is using another trampoline for another part of the payment?

Reply:
> it may aggregate HTLCs before forwarding, but what does "aggregate" mean in that context? simply to wait? what if the sender is using another trampoline for another part of the payment?

Actually the spec is quite confusing right now, my bad, I tried to keep this part concise and kinda messed it up.

If the trampoline node receives an HTLC which is a part of an MPP (because `total_amount_msat` > `htlc_amount` in the *outer* onion), it **MUST** aggregate it (ie wait to receive all the parts that he's supposed to receive) before forwarding. Otherwise we would not be able to forward the requested outgoing amount, because he could have received less than that!

An example will help clarify this. Alice sends a payment to Carol through Bob (which is a trampoline node).
Alice decides to send `10,000 sats` through Bob (and may send more sats for the same payment through another trampoline node). But Alice doesn't have a route to Bob that can carry `10,000 sats` in a single HTLC, so she splits it in two HTLCs (for example `4,000 sats` and `6,500 sats` for a total of `10,500 sats` because Bob takes a `500 sats` relay fee).

When Bob receives the first HTLC, the outer onion tells it that the `total_amount = 10,500 sats`. The inner onion tells it to relay `10,000 sat` to Carol. So Bob must wait to receive the full `10,500 sats` before forwarding the payment to Carol. That's what I call "aggregate" here.

If the sender doesn't want the trampoline node to aggregate the payments, the sender must set each HTLC's `total_amount` to equal the `htlc_amount`, and must use a different inner onion for each htlc (with a different `outgoing_amount` for each HTLC). This will likely cost more fees and be less efficient, but is supported.

Reply:
Yes, that explanation is clear. By "the trampoline is the final recipient of the MPP", we both mean `total_amount_msat > htlc_amount` in the outer onion. That case is indeed very clear: the trampoline must aggregate. My question was about the case where the trampoline is not the recipient of the MPP. I do not think that "aggregate" makes sense in that case.

In other words, "MAY aggregate" is confusing, because it gives the feeling that the decision to aggregate is with the trampoline node.

Reply:
> My question was about the case where the trampoline is not the recipient of the MPP. I do not think that "aggregate" makes sense in that case.

You're right, in that case the trampoline even must **not** aggregate because it doesn't know how much it should even receive before forwarding. As you say, that decision is indeed instead made by the sender (not the trampoline node) depending on what `total_amount_msat` it puts in the outer onion.

Reply:
I've clarified this in https://github.com/lightning/bolts/pull/836/commits/985cf69e25d81174545048bd70790b8a33605118